### DISCUSSION TITLE
Is there any reason why I cannot deconstruct dynamic values into typed variables?

### DISCUSSION THREAD

**Discussion by jshirota** [dotnet/csharplang#5462] [Q&A]
Is there any reason why I cannot deconstruct dynamic values into typed variables?

```cs

using System;

var cat = new Cat();

// This is OK.
(dynamic x, dynamic y) = cat;
var name = (string)x;
var age = (int)y;

// This is not.  Why?
// (string name, int age) = cat;

Console.WriteLine($"{name} {age}");

public class Cat
{
    // Some dynamic data from somewhere.
    private dynamic[] data = { "Leo", 8 };

    public void Deconstruct(out dynamic name, out dynamic age)
    {
        name = data[0];
        age = data[1];
    }
}

```

_Originally posted by @jshirota in https://github.com/dotnet/csharplang/discussions/4892_

**Comment by 333fred** [comment]
Think of the assignment as turning into `cat.Deconstruct(out Type name, out Type age)`. If that doesn't work (it doesn't here), then the deconstruction assignment won't either.

**Reply by jshirota** [reply]
So this boils down to this?

```cs
// OK.
GetNumber(out var n);

// Not OK.  Why?
// GetNumber(out int n);

void GetNumber(out dynamic n) => n = 123;
```

**Reply by 333fred** [reply]
That's correct. And the reason that doesn't work is because you have a local of type `int`, not of type object (`dynamic` is `object` + compiler magic).

**Reply by jnm2** [reply]
I guess you could say compiler magic, but also specifically not enough compiler magic to hide an explicit cast to an `int` variable.

**Reply by FaustVX** [reply]
@333fred
I always thought `dynamic` could be assigned to/from anything. And if something can't work, it fails a runtime, not at compile time.

**Reply by svick** [reply]
@333fred I don't think that answer is satisfying. For example if you have a class `C` with `void Deconstruct(out int i, out int j)`, you can do `(long x, long y) = c;`, even though `c.Deconstruct(out long x, out long y);` won't compile. [See SharpLab.](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8BuEyABAMbEC8xMApgO7EDCAFAJQDcAsAFDMA2AexgBzYgA84xQSOIBPVpTJdupAHQARGqSEBnBGlIJmA1AilDRESSbPTRwDjx4ABAExMeAbx7FfxZwDM/gAsxJraMHoGRjbEsGZQ1qZxMGYAVgoUAHzEzInEGYrMAIxwro7cAL5AA==)

**Reply by 333fred** [reply]
`int`->`long` is an implicit numeric conversion, and is a standard conversion. `dynamic`->`int` is _not_ a standard conversion, and therefore cannot have the same substitution applied.

**Reply by jnm2** [reply]
Yeah, I want to know the answer to @FaustVX's point too. Why is it that assignment and `out` don't convert the same way? Is it for the same reason that C# doesn't allow properties to be passed to `ref`/`out` parameters, where an implicit temporary variable would have to be introduced?

**Reply by FaustVX** [reply]
@jnm2 
[[On Sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AoA2B7GBzAAgA84Cd8CBPASgIF4CYBTAdwIGF1qBuAWACgBAAQBMHAQG8BBGQSEBmOQBYCAESYBjXAGcALgCcArht3psh3QViWopczZiWAVtWmz3dAHwF0tgi/ofAEY4ER4BAF8gA==)], the implicit call to `Deconstruct` already introduce temporary variable.

**Reply by jshirota** [reply]
I really don't see any consistent pattern here from the user's point of view.

```cs
C1.GetNumber(out long a);
C1.GetNumber(out dynamic b);
C2.GetNumber(out long c);
C2.GetNumber(out dynamic d);        // OK.
new C3().GetNumber(out long e);
new C3().GetNumber(out dynamic f);
new C4().GetNumber(out long g);
new C4().GetNumber(out dynamic h);  // OK.
(long i, _) = new C5();             // OK.
(dynamic j, _) = new C5();          // OK.
(long k, _) = new C6();
(dynamic l, _) = new C6();          // OK.

class C1 { public static void GetNumber(out int x) => x = 1; }
class C2 { public static void GetNumber(out dynamic x) => x = 1; }
class C3 { public void GetNumber(out int x) => x = 1; }
class C4 { public void GetNumber(out dynamic x) => x = 1; }
class C5 { public void Deconstruct(out int x, out int y) => (x, y) = (1, 2); }
class C6 { public void Deconstruct(out dynamic x, out dynamic y) => (x, y) = (1, 2); }
```
